### PR TITLE
Doc: Add JSON WebService example

### LINKED ISSUE #217: Add an example that powers a Json based web-service

### PR DESCRIPTION
**Pull request by ShrutiVerma97**
fixes #217

### CODE REVIEW

File: example/src/main/scala/JsonWebService.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+object JsonWebService extends App {
+
+  case class Employee(id: String, name: String, Experience: String)
+  val employees = List(Employee("1", "abc", "1yr"), Employee("2", "def", "2yrs"), Employee("3", "xyz", "3yrs"))
+
+  //get employee details if employee exists
+  def getDetails(id: String): String = {
+    val emp = employees.filter(_.id.equals(id))
+    if (emp.isEmpty) s"""Employee doesn't exist"""
+    else s"""{"id": "${emp.head.id}", "name": "${emp.head.name}", "experience": "${emp.head.Experience}"}"""
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala:17]: I think we should use `zio-json` for this example instead of hard coding.

Comment (ShrutiVerma97) [reply example/src/main/scala/JsonWebService.scala]: @tusharmath zio-json lacks Scala3 support: https://github.com/zio/zio-json/issues/15, that's why the build is failing. Should I use something else?

File: example/src/main/scala/JsonWebService.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  implicit val employeeEncoder: Encoder[Employee] =
+    Encoder.forProduct3("id", "name", "experience")(emp => (emp.id, emp.name, emp.experience))
+
+  val employees = List(Employee(1, "abc", 3), Employee(2, "def", 2), Employee(3, "xyz", 4))
+
+  //get employee details if employee exists
+  def getDetails(id: String): String = {
+    val emp = employees.filter(_.id.toString.equals(id))
+    if (emp.isEmpty) s"""Employee doesn't exist"""
+    else emp.head.asJson.noSpaces
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala:23]:
```suggestion
    employees.filter(_.id.toString.equals(id)).asJson.noSpaces
```

File: example/src/main/scala/JsonWebService.scala [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  // Create HTTP route
+  val app: HttpApp[Any, Nothing] = HttpApp.collect { case Method.GET -> Root / "get-employee-details" / id =>
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala:27]:
```suggestion
  val app: HttpApp[Any, Nothing] = HttpApp.collect { case Method.GET -> Root / "employee" / id =>
```

File: example/src/main/scala/JsonWebService.scala [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Example to build app using JSON web service
+ */
+object JsonWebService extends App {
+
+  final case class Employee(id: Long, name: String, experience: Int)
+
+  implicit val employeeEncoder: Encoder[Employee] =
+    Encoder.forProduct3("id", "name", "experience")(emp => (emp.id, emp.name, emp.experience))
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala:15]: Use circe to generate encoder.

File: example/src/main/scala/JsonWebService.scala [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+  val employees = List(Employee(1, "abc", 3), Employee(2, "def", 2), Employee(3, "xyz", 4))
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala:17]: Use `zio.test.Gen` to randomly generate users.

File: example/src/main/scala/JsonWebService.scala [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  //get employee details if employee exists
+  def getDetails(id: String): String = {
+    val emp = employees.filter(_.id.toString.equals(id))
+    if (emp.isEmpty) s"""Employee doesn't exist"""
+    else emp.head.asJson.noSpaces
+  }
+
+  // Create HTTP route
+  val app: HttpApp[Any, Nothing] = HttpApp.collect { case Method.GET -> Root / "get-employee-details" / id =>
+    Response.jsonString(getDetails(id))
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala:28]: Add another route to create users. 

Final app can look like `Http[R, E, UserRequest, UserResponse]`
where both `UserRequest` and `UserResponse` are sum types

File: example/src/main/scala/JsonWebService.scala [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    case class Get(req: Request)                           extends UserRequest
+    case object BadRequest                                 extends UserRequest
+  }
+  sealed trait UserResponse
+  object UserResponse {
+    case class StatusResponse(msg: String) extends UserResponse
+    case class JsonResponse(data: String)  extends UserResponse
+  }
+
+  final case class Employee(id: Long, name: String, experience: Int)
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]:
```suggestion
  final case class User(id: Long, name: String, experience: Int)
```

File: example/src/main/scala/JsonWebService.scala [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import zio.test.{Gen, Sized}
+
+/**
+ * Example to build app using JSON web service
+ */
+object JsonWebService extends App {
+
+  sealed trait UserRequest
+  object UserRequest  {
+    case class Post(req: Request, content: Option[String]) extends UserRequest
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]:
```suggestion
    case class Create(user: User) extends UserRequest
```

File: example/src/main/scala/JsonWebService.scala [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    case class Get(req: Request)                           extends UserRequest
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]:
```suggestion
    case object Show                           extends UserRequest
```

File: example/src/main/scala/JsonWebService.scala [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    case object BadRequest                                 extends UserRequest
+  }
+  sealed trait UserResponse
+  object UserResponse {
+    case class StatusResponse(msg: String) extends UserResponse
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]:
```suggestion
    case class Show(users: List[User]) extends UserResponse
```

File: example/src/main/scala/JsonWebService.scala [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    case class JsonResponse(data: String)  extends UserResponse
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]:
```suggestion
    case class Create(user: User)  extends UserResponse
```

File: example/src/main/scala/JsonWebService.scala [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  final case class User(id: Long, name: String)
+
+  def getUser(jsonString: String): User = {
+    val json: Json      = parse(jsonString).getOrElse(null)
+    val cursor: HCursor = json.hcursor
+    val id              = cursor.downField("id").as[Long]
+    val name            = cursor.downField("name").as[String]
+    User(id.getOrElse(0), name.getOrElse(""))
+  }
+  var users                                                  = List.empty[User]
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]: Use `Ref[List[User]]`

File: example/src/main/scala/JsonWebService.scala [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    Http.collectM[UserRequest]({
+      case UserRequest.Show         => UIO(UserResponse.Show(users))
+      case UserRequest.Create(user) => {
+        users = users :+ user
+        UIO(UserResponse.Create(user))
+      }
+    })
+
+  // Create HTTP route
+  val app: Http[Any, Nothing, Request, UResponse] = userApp
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]:
```suggestion
  def app(ref: Ref[List[User]]): Http[Any, Nothing, Request, UResponse] = userApp
```

File: example/src/main/scala/JsonWebService.scala [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    .map {
+      case UserResponse.Show(users)  => Response.text(users.asJson.noSpaces)
+      case UserResponse.Create(user) => Response.text(user.asJson.noSpaces)
+    }
+
+  // Run it like any simple app
+  override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] = {
+    val ref: UIO[Ref[List[User]]] = Ref.make(List[User]())
+    for {
+      r <- ref
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]: can inline the Ref creation here.

File: example/src/main/scala/JsonWebService.scala [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+import zio._
+
+/**
+ * Example to build app using JSON web service
+ */
+object JsonWebService extends App {
+
+  sealed trait UserRequest
+  object UserRequest  {
+    case class Create(user: User, ref: Ref[List[User]]) extends UserRequest
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]: `ref` isn't needed.

File: example/src/main/scala/JsonWebService.scala [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    case class Show(ref: Ref[List[User]])               extends UserRequest
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]: `ref` isn't needed.

File: example/src/main/scala/JsonWebService.scala [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+  }
+  sealed trait UserResponse
+  object UserResponse {
+    case class Show(users: List[User]) extends UserResponse
+    case class Create(user: User)      extends UserResponse
+  }
+
+  final case class User(id: Long, name: String)
+
+  def getUser(jsonString: String): User = {
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]: Use circe's inbuilt codecs.

File: example/src/main/scala/JsonWebService.scala [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+    val json: Json      = parse(jsonString).getOrElse(null)
+    val cursor: HCursor = json.hcursor
+    val id              = cursor.downField("id").as[Long]
+    val name            = cursor.downField("name").as[String]
+    User(id.getOrElse(0), name.getOrElse(""))
+  }
+  val userApp: Http[Any, Nothing, UserRequest, UserResponse] =
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]: convert to def and pass the one ref that is created in the begining.

File: example/src/main/scala/JsonWebService.scala [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+    Http.collectM[UserRequest]({
+      case UserRequest.Show         => {
+        for {
+          users <- ref.get
+        } yield UserResponse.Show(users)
+      }
+      case UserRequest.Create(user) => {
+        for {
+          users <- ref.get
+          _     <- ref.set(users :+ user)
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]: can use `ref.update`

File: example/src/main/scala/JsonWebService.scala [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+    case object BadRequest        extends UserRequest
+  }
+  sealed trait UserResponse
+  object UserResponse {
+    case class Show(users: List[User]) extends UserResponse
+    case class Create(user: User)      extends UserResponse
+  }
+  final case class User(id: Long, name: String)
+
+  def getUser(jsonString: String): User = decode[User](jsonString).getOrElse(null)
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]: Return an `Either`. Send a `400` status code if there is a decoding failure

File: example/src/main/scala/JsonWebService.scala [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * Example to build app using JSON web service
+ */
+object JsonWebService extends App {
+
+  sealed trait UserRequest
+  object UserRequest  {
+    case class Create(user: User) extends UserRequest
+    case object Show              extends UserRequest
+    case object BadRequest        extends UserRequest
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]: Might not be needed.

File: example/src/main/scala/JsonWebService.scala [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+  }
+  sealed trait UserResponse
+  object UserResponse {
+    case class Show(users: List[User]) extends UserResponse
+    case class Create(user: User)      extends UserResponse
+    case object BadResponse            extends UserResponse
```
Comment (tusharmath) [inline example/src/main/scala/JsonWebService.scala]: Might not be needed. Can use `Http.fail` for this case.

**Pull request closed without merging**